### PR TITLE
- rename BufferAndTypeId -> BufferAndMeta and add the array size to it.

### PR DESCRIPTION
**Pull request by baldersheim**
- Use _buffers instead of _states to get array size.

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: searchlib/src/vespa/searchlib/memoryindex/feature_store.h [hunk 1]
```diff
@@ -154,12 +154,8 @@ class FeatureStore {
...
         const uint8_t * bits = getBits(ref);
         decoder.setByteCompr(bits);
-        uint32_t bufferId = RefType(ref).bufferId();
-        const vespalib::datastore::BufferState &state = _store.getBufferState(bufferId);
-        decoder.setEnd(
-                ((_store.getEntryArray<uint8_t>(RefType(0, bufferId), buffer_array_size) + state.size() -
-                  bits) + 7) / 8,
-                false);
+        constexpr uint32_t maxOffset = RefType::offsetSize() * buffer_array_size;
+        decoder.setEnd(maxOffset, false);
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/memoryindex/feature_store.h:158]: @toregge Can you please review this change?

File: vespalib/src/vespa/vespalib/datastore/bufferstate.h [hunk 2]
```diff
@@ -130,7 +130,27 @@ class BufferState
     BufferTypeBase *getTypeHandler() { return _typeHandler.load(std::memory_order_relaxed); }
 
     void resume_primary_buffer(uint32_t buffer_id);
+};
 
+class BufferAndMeta {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/bufferstate.h:135]: Consider moving this class to a separate header file.

**Review by geirst — APPROVED** [review]
👍  @toregge please review as well

File: vespalib/src/vespa/vespalib/datastore/bufferstate.h (hunk 2 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/datastore/bufferstate.h:135]: I prefer keeping it here as it is a small subset of the BufferState.

**Review by toregge — APPROVED** [review]
:+1:

File: vespalib/src/vespa/vespalib/datastore/bufferstate.h (hunk 2 above)
Comment (geirst) [reply vespalib/src/vespa/vespalib/datastore/bufferstate.h:135]: 👍

**Pull request merged**